Error while checking build/O2/fullCI for f75ba9f4a966c8750e2cbc0cb786c92ab0b45e18 at 2023-11-24 16:02:
```
## sw/BUILD/O2-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
/sw/SOURCES/O2/11825-slc8_x86-64/0/Common/Utils/src/ConfigurableParamReaders.cxx:56:25: error: 'exists' is not a member of 'std::filesystem'
ninja: build stopped: subcommand failed.
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/11825/f75ba9f4a966c8750e2cbc0cb786c92ab0b45e18/build_O2_fullCI/pretty.html).

--- a/Common/SimConfig/test/testSimCutParam.cxx
+++ b/Common/SimConfig/test/testSimCutParam.cxx
@@ -15,6 +15,7 @@
 #include <boost/test/unit_test.hpp>
 #include "SimConfig/SimParams.h"
 #include "CommonUtils/ConfigurableParam.h"
+#include <boost/property_tree/ptree.hpp>
 
 using namespace o2::conf;
 

--- a/Common/Utils/CMakeLists.txt
+++ b/Common/Utils/CMakeLists.txt
@@ -14,7 +14,7 @@ o2_add_library(CommonUtils
                        src/RootChain.cxx src/CompStream.cxx src/ShmManager.cxx
                        src/ValueMonitor.cxx
                        src/StringUtils.cxx
-                       src/ConfigurableParamHelper.cxx src/ConfigurableParam.cxx src/RootSerializableKeyValueStore.cxx
+                       src/ConfigurableParamReaders.cxx src/ConfigurableParamHelper.cxx src/ConfigurableParam.cxx src/RootSerializableKeyValueStore.cxx
                        src/KeyValParam.cxx
                        src/FileSystemUtils.cxx
                        src/FIFO.cxx
@@ -39,6 +39,7 @@ o2_target_root_dictionary(CommonUtils
                                   include/CommonUtils/MemFileHelper.h
                                   include/CommonUtils/ConfigurableParam.h
                                   include/CommonUtils/ConfigurableParamHelper.h
+                                  include/CommonUtils/ConfigurableParamReaders.h
                                   include/CommonUtils/ConfigurationMacroHelper.h
                                   include/CommonUtils/RootSerializableKeyValueStore.h
                                   include/CommonUtils/KeyValParam.h

--- a/Common/Utils/include/CommonUtils/ConfigurableParam.h
+++ b/Common/Utils/include/CommonUtils/ConfigurableParam.h
@@ -18,7 +18,7 @@
 #include <cassert>
 #include <map>
 #include <unordered_map>
-#include <boost/property_tree/ptree.hpp>
+#include <boost/property_tree/ptree_fwd.hpp>
 #include <typeinfo>
 #include <iostream>
 #include <array>
@@ -176,15 +176,10 @@ class ConfigurableParam
   static void printAllRegisteredParamNames();
   static void printAllKeyValuePairs(bool useLogger = false);
 
-  static const std::string& getInputDir() { return sInputDir; }
   static const std::string& getOutputDir() { return sOutputDir; }
 
-  static void setInputDir(const std::string& d) { sInputDir = d; }
   static void setOutputDir(const std::string& d) { sOutputDir = d; }
 
-  static boost::property_tree::ptree readINI(std::string const& filepath);
-  static boost::property_tree::ptree readJSON(std::string const& filepath);
-  static boost::property_tree::ptree readConfigFile(std::string const& filepath);
   static bool configFileExists(std::string const& filepath);
 
   // writes a human readable JSON file of all parameters
@@ -196,10 +191,12 @@ class ConfigurableParam
   template <typename T>
   static T getValueAs(std::string key)
   {
-    if (!sIsFullyInitialized) {
-      initialize();
-    }
-    return sPtree->get<T>(key);
+    return [](auto* tree, const std::string& key) -> T {
+      if (!sIsFullyInitialized) {
+        initialize();
+      }
+      return tree->template get<T>(key);
+    }(sPtree, key);
   }
 
   template <typename T>
@@ -208,19 +205,21 @@ class ConfigurableParam
     if (!sIsFullyInitialized) {
       initialize();
     }
-    assert(sPtree);
-    try {
-      auto key = mainkey + "." + subkey;
-      if (sPtree->get_optional<std::string>(key).is_initialized()) {
-        sPtree->put(key, x);
-        auto changed = updateThroughStorageMap(mainkey, subkey, typeid(T), (void*)&x);
-        if (changed != EParamUpdateStatus::Failed) {
-          sValueProvenanceMap->find(key)->second = kRT; // set to runtime
+    return [&subkey, &x, &mainkey](auto* tree) -> void {
+      assert(tree);
+      try {
+        auto key = mainkey + "." + subkey;
+        if (tree->template get_optional<std::string>(key).is_initialized()) {
+          tree->put(key, x);
+          auto changed = updateThroughStorageMap(mainkey, subkey, typeid(T), (void*)&x);
+          if (changed != EParamUpdateStatus::Failed) {
+            sValueProvenanceMap->find(key)->second = kRT; // set to runtime
+          }
         }
+      } catch (std::exception const& e) {
+        std::cerr << "Error in setValue (T) " << e.what() << "\n";
       }
-    } catch (std::exception const& e) {
-      std::cerr << "Error in setValue (T) " << e.what() << "\n";
-    }
+    }(sPtree);
   }
 
   static void setProvenance(std::string const& mainkey, std::string const& subkey, EParamProvenance p)
@@ -242,25 +241,7 @@ class ConfigurableParam
 
   // specialized for std::string
   // which means that the type will be converted internally
-  static void setValue(std::string const& key, std::string const& valuestring)
-  {
-    if (!sIsFullyInitialized) {
-      initialize();
-    }
-    assert(sPtree);
-    try {
-      if (sPtree->get_optional<std::string>(key).is_initialized()) {
-        sPtree->put(key, valuestring);
-        auto changed = updateThroughStorageMapWithConversion(key, valuestring);
-        if (changed != EParamUpdateStatus::Failed) {
-          sValueProvenanceMap->find(key)->second = kRT; // set to runtime
-        }
-      }
-    } catch (std::exception const& e) {
-      std::cerr << "Error in setValue (string) " << e.what() << "\n";
-    }
-  }
-
+  static void setValue(std::string const& key, std::string const& valuestring);
   static void setEnumValue(const std::string&, const std::string&);
   static void setArrayValue(const std::string&, const std::string&);
 
@@ -321,7 +302,6 @@ class ConfigurableParam
   // (stored as a vector of pairs <enumValueLabel, enumValueInt>)
   static EnumRegistry* sEnumRegistry;
 
-  static std::string sInputDir;
   static std::string sOutputDir;
 
   void setRegisterMode(bool b) { sRegisterMode = b; }

--- a/Common/Utils/include/CommonUtils/ConfigurableParamReaders.h
+++ b/Common/Utils/include/CommonUtils/ConfigurableParamReaders.h
@@ -1,0 +1,36 @@
+// Copyright 2019-2020 CERN and copyright holders of ALICE O2.
+// See https://alice-o2.web.cern.ch/copyright for details of the copyright holders.
+// All rights not expressly granted are reserved.
+//
+// This software is distributed under the terms of the GNU General Public
+// License v3 (GPL Version 3), copied verbatim in the file "COPYING".
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+#ifndef O2_COMMON_UTILS_CONFIGURABLEPARAMREADERS_H_
+#define O2_COMMON_UTILS_CONFIGURABLEPARAMREADERS_H_
+
+#include <boost/property_tree/ptree.hpp>
+#include <string>
+
+namespace o2::conf
+{
+
+// Helpers to read ConfigurableParam from different file formats
+class ConfigurableParamReaders
+{
+ public:
+  static void setInputDir(const std::string& d) { sInputDir = d; }
+  static const std::string& getInputDir() { return sInputDir; }
+
+  static boost::property_tree::ptree readINI(std::string const& filepath);
+  static boost::property_tree::ptree readJSON(std::string const& filepath);
+  static boost::property_tree::ptree readConfigFile(std::string const& filepath);
+
+ private:
+  static std::string sInputDir;
+};
+
+} // namespace o2::conf
+#endif // O2_COMMON_UTILS_CONF_CONFIGURABLEPARAMREADERS_H_

--- a/Common/Utils/src/ConfigurableParam.cxx
+++ b/Common/Utils/src/ConfigurableParam.cxx
@@ -9,11 +9,12 @@
 // granted to it by virtue of its status as an Intergovernmental Organization
 // or submit itself to any jurisdiction.
 
-//first version 8/2018, Sandro Wenzel
+// first version 8/2018, Sandro Wenzel
 
 #include "CommonUtils/ConfigurableParam.h"
 #include "CommonUtils/StringUtils.h"
 #include "CommonUtils/KeyValParam.h"
+#include "CommonUtils/ConfigurableParamReaders.h"
 #define BOOST_BIND_GLOBAL_PLACEHOLDERS
 #include <boost/algorithm/string/predicate.hpp>
 #include <boost/property_tree/ptree.hpp>
@@ -47,7 +48,6 @@ std::vector<ConfigurableParam*>* ConfigurableParam::sRegisteredParamClasses = nu
 boost::property_tree::ptree* ConfigurableParam::sPtree = nullptr;
 std::map<std::string, std::pair<std::type_info const&, void*>>* ConfigurableParam::sKeyToStorageMap = nullptr;
 std::map<std::string, ConfigurableParam::EParamProvenance>* ConfigurableParam::sValueProvenanceMap = nullptr;
-std::string ConfigurableParam::sInputDir = "";
 std::string ConfigurableParam::sOutputDir = "";
 EnumRegistry* ConfigurableParam::sEnumRegistry = nullptr;
 
@@ -194,60 +194,28 @@ void ConfigurableParam::writeINI(std::string const& filename, std::string const&
 
 bool ConfigurableParam::configFileExists(std::string const& filepath)
 {
-  return std::filesystem::exists(o2::utils::Str::concat_string(sInputDir, filepath));
+  return std::filesystem::exists(o2::utils::Str::concat_string(ConfigurableParamReaders::getInputDir(), filepath));
 }
 
 // ------------------------------------------------------------------
 
-boost::property_tree::ptree ConfigurableParam::readConfigFile(std::string const& filepath)
+void ConfigurableParam::setValue(std::string const& key, std::string const& valuestring)
 {
-  auto inpfilename = o2::utils::Str::concat_string(sInputDir, filepath);
-  if (!std::filesystem::exists(inpfilename)) {
-    LOG(fatal) << inpfilename << " : config file does not exist!";
+  if (!sIsFullyInitialized) {
+    initialize();
   }
-
-  boost::property_tree::ptree pt;
-
-  if (boost::iends_with(inpfilename, ".ini")) {
-    pt = readINI(inpfilename);
-  } else if (boost::iends_with(inpfilename, ".json")) {
-    pt = readJSON(inpfilename);
-  } else {
-    LOG(fatal) << "Configuration file must have either .ini or .json extension";
-  }
-
-  return pt;
-}
-
-// ------------------------------------------------------------------
-
-boost::property_tree::ptree ConfigurableParam::readINI(std::string const& filepath)
-{
-  boost::property_tree::ptree pt;
+  assert(sPtree);
   try {
-    boost::property_tree::read_ini(filepath, pt);
-  } catch (const boost::property_tree::ptree_error& e) {
-    LOG(fatal) << "Failed to read INI config file " << filepath << " (" << e.what() << ")";
-  } catch (...) {
-    LOG(fatal) << "Unknown error when reading INI config file ";
+    if (sPtree->get_optional<std::string>(key).is_initialized()) {
+      sPtree->put(key, valuestring);
+      auto changed = updateThroughStorageMapWithConversion(key, valuestring);
+      if (changed != EParamUpdateStatus::Failed) {
+        sValueProvenanceMap->find(key)->second = kRT; // set to runtime
+      }
+    }
+  } catch (std::exception const& e) {
+    std::cerr << "Error in setValue (string) " << e.what() << "\n";
   }
-
-  return pt;
-}
-
-// ------------------------------------------------------------------
-
-boost::property_tree::ptree ConfigurableParam::readJSON(std::string const& filepath)
-{
-  boost::property_tree::ptree pt;
-
-  try {
-    boost::property_tree::read_json(filepath, pt);
-  } catch (const boost::property_tree::ptree_error& e) {
-    LOG(fatal) << "Failed to read JSON config file " << filepath << " (" << e.what() << ")";
-  }
-
-  return pt;
 }
 
 // ------------------------------------------------------------------
@@ -258,7 +226,7 @@ void ConfigurableParam::writeJSON(std::string const& filename, std::string const
     LOG(info) << "ignoring writing of json file " << filename;
     return;
   }
-  initPropertyTree();     // update the boost tree before writing
+  initPropertyTree(); // update the boost tree before writing
   auto outfilename = o2::utils::Str::concat_string(sOutputDir, filename);
   if (!keyOnly.empty()) { // write ini for selected key only
     try {
@@ -409,7 +377,7 @@ void ConfigurableParam::updateFromFile(std::string const& configFile, std::strin
     return;
   }
 
-  boost::property_tree::ptree pt = readConfigFile(cfgfile);
+  boost::property_tree::ptree pt = ConfigurableParamReaders::readConfigFile(cfgfile);
 
   std::vector<std::pair<std::string, std::string>> keyValPairs;
   auto request = o2::utils::Str::tokenize(paramsList, ',', true);
@@ -522,7 +490,7 @@ void ConfigurableParam::updateFromString(std::string const& configString)
 
   const auto& kv = o2::conf::KeyValParam::Instance();
   if (getProvenance("keyval.input_dir") != kCODE) {
-    sInputDir = o2::utils::Str::concat_string(o2::utils::Str::rectifyDirectory(kv.input_dir));
+    ConfigurableParamReaders::setInputDir(o2::utils::Str::concat_string(o2::utils::Str::rectifyDirectory(kv.input_dir)));
   }
   if (getProvenance("keyval.output_dir") != kCODE) {
     if (kv.output_dir == "/dev/null") {

--- a/Common/Utils/src/ConfigurableParamReaders.cxx
+++ b/Common/Utils/src/ConfigurableParamReaders.cxx
@@ -1,0 +1,76 @@
+// Copyright 2019-2020 CERN and copyright holders of ALICE O2.
+// See https://alice-o2.web.cern.ch/copyright for details of the copyright holders.
+// All rights not expressly granted are reserved.
+//
+// This software is distributed under the terms of the GNU General Public
+// License v3 (GPL Version 3), copied verbatim in the file "COPYING".
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+#include "CommonUtils/ConfigurableParamReaders.h"
+#include "CommonUtils/StringUtils.h"
+#include <fairlogger/Logger.h>
+#include <filesystem>
+
+#include <boost/algorithm/string/predicate.hpp>
+#include <boost/property_tree/ini_parser.hpp>
+#include <boost/property_tree/json_parser.hpp>
+
+namespace o2::conf
+{
+// ------------------------------------------------------------------
+
+boost::property_tree::ptree ConfigurableParamReaders::readINI(std::string const& filepath)
+{
+  boost::property_tree::ptree pt;
+  try {
+    boost::property_tree::read_ini(filepath, pt);
+  } catch (const boost::property_tree::ptree_error& e) {
+    LOG(fatal) << "Failed to read INI config file " << filepath << " (" << e.what() << ")";
+  } catch (...) {
+    LOG(fatal) << "Unknown error when reading INI config file ";
+  }
+
+  return pt;
+}
+
+// ------------------------------------------------------------------
+
+boost::property_tree::ptree ConfigurableParamReaders::readJSON(std::string const& filepath)
+{
+  boost::property_tree::ptree pt;
+
+  try {
+    boost::property_tree::read_json(filepath, pt);
+  } catch (const boost::property_tree::ptree_error& e) {
+    LOG(fatal) << "Failed to read JSON config file " << filepath << " (" << e.what() << ")";
+  }
+
+  return pt;
+}
+
+boost::property_tree::ptree ConfigurableParamReaders::readConfigFile(std::string const& filepath)
+{
+  auto inpfilename = o2::utils::Str::concat_string(sInputDir, filepath);
+  if (!std::filesystem::exists(inpfilename)) {
+    LOG(fatal) << inpfilename << " : config file does not exist!";
+  }
+
+  boost::property_tree::ptree pt;
+
+  if (boost::iends_with(inpfilename, ".ini")) {
+    pt = ConfigurableParamReaders::readINI(inpfilename);
+  } else if (boost::iends_with(inpfilename, ".json")) {
+    pt = ConfigurableParamReaders::readJSON(inpfilename);
+  } else {
+    LOG(fatal) << "Configuration file must have either .ini or .json extension";
+  }
+
+  return pt;
+}
+
+std::string ConfigurableParamReaders::sInputDir = "";
+
+} // namespace o2::conf

--- a/Detectors/MUON/MCH/Raw/test/testClosureCoDec.cxx
+++ b/Detectors/MUON/MCH/Raw/test/testClosureCoDec.cxx
@@ -16,6 +16,7 @@
 #define BOOST_TEST_DYN_LINK
 
 #include <boost/test/unit_test.hpp>
+#include <boost/property_tree/ptree.hpp>
 #include "DetectorsRaw/HBFUtils.h"
 #include "DetectorsRaw/RawFileWriter.h"
 #include "Framework/Logger.h"

--- a/Detectors/MUON/MCH/Simulation/test/testDigitizer.cxx
+++ b/Detectors/MUON/MCH/Simulation/test/testDigitizer.cxx
@@ -31,6 +31,7 @@
 #include "TGeoManager.h"
 #include "boost/format.hpp"
 #include <boost/test/data/test_case.hpp>
+#include <boost/property_tree/ptree.hpp>
 #include <algorithm>
 #include <unordered_map>
 

--- a/Detectors/MUON/MCH/Triggering/test/testEventFinder.cxx
+++ b/Detectors/MUON/MCH/Triggering/test/testEventFinder.cxx
@@ -27,6 +27,7 @@
 #include <vector>
 
 #include <boost/test/unit_test.hpp>
+#include <boost/property_tree/ptree.hpp>
 
 #include "CommonUtils/ConfigurableParam.h"
 #include "SimulationDataFormat/MCCompLabel.h"

--- a/Detectors/TPC/base/test/testTPCParameters.cxx
+++ b/Detectors/TPC/base/test/testTPCParameters.cxx
@@ -23,9 +23,9 @@
 #include "TPCBase/ParameterGas.h"
 #include <CommonUtils/ConfigurableParam.h>
 #include <CommonUtils/ConfigurableParamHelper.h>
-namespace o2
-{
-namespace tpc
+#include <boost/property_tree/ptree.hpp>
+
+namespace o2::tpc
 {
 
 constexpr float NominalTimeBin = 8 * o2::constants::lhc::LHCBunchSpacingNS * 1e-3;
@@ -367,5 +367,4 @@ BOOST_AUTO_TEST_CASE(ParameterGEM_test2)
   BOOST_CHECK_CLOSE(o2::conf::ConfigurableParam::getValueAs<float>("TPCGEMParam.KappaStack"), 32, 1e-12);
   BOOST_CHECK_CLOSE(o2::conf::ConfigurableParam::getValueAs<float>("TPCGEMParam.EfficiencyStack"), 33, 1e-3);
 }
-} // namespace tpc
 } // namespace o2

--- a/Detectors/TPC/calibration/include/TPCCalibration/IDCAverageGroup.h
+++ b/Detectors/TPC/calibration/include/TPCCalibration/IDCAverageGroup.h
@@ -20,6 +20,7 @@
 #include "TPCCalibration/IDCAverageGroupBase.h"
 #include "TPCBase/Sector.h"
 #include "TPCBase/CalDet.h"
+#include <boost/property_tree/ptree.hpp>
 #include <vector>
 #include "Rtypes.h"
 

--- a/Detectors/TPC/calibration/include/TPCCalibration/IDCFactorization.h
+++ b/Detectors/TPC/calibration/include/TPCCalibration/IDCFactorization.h
@@ -24,6 +24,7 @@
 #include "TPCCalibration/IDCContainer.h"
 #include "TPCCalibration/IDCGroupHelperSector.h"
 #include "DataFormatsTPC/Defs.h"
+#include <boost/property_tree/ptree.hpp>
 
 namespace o2::tpc
 {

--- a/Detectors/TPC/calibration/include/TPCCalibration/SACFactorization.h
+++ b/Detectors/TPC/calibration/include/TPCCalibration/SACFactorization.h
@@ -21,6 +21,7 @@
 #include "Rtypes.h"
 #include "TPCCalibration/IDCContainer.h"
 #include "DataFormatsTPC/Defs.h"
+#include <boost/property_tree/ptree.hpp>
 
 namespace o2::tpc
 {

--- a/Detectors/TPC/calibration/src/IDCGroupingParameter.cxx
+++ b/Detectors/TPC/calibration/src/IDCGroupingParameter.cxx
@@ -12,6 +12,8 @@
 #include "TPCCalibration/IDCGroupingParameter.h"
 #include "Framework/Logger.h"
 #include "Algorithm/RangeTokenizer.h"
+#include <boost/property_tree/ptree.hpp>
+#include <cassert>
 #include <stdexcept>
 
 using namespace o2::tpc;

--- a/Detectors/TPC/reconstruction/macro/makeIonTail.C
+++ b/Detectors/TPC/reconstruction/macro/makeIonTail.C
@@ -32,6 +32,7 @@
 #include "TPCBase/CRUCalibHelpers.h"
 #include "TPCBase/CRU.h"
 #include "CommonUtils/TreeStreamRedirector.h"
+#include <boost/property_tree/ptree.hpp>
 
 using namespace o2::tpc;
 size_t digitsInSaturateion(std::vector<Digit>& digits, bool correctCharge = false, CalPad* pedestals = nullptr, o2::utils::TreeStreamRedirector* stream = nullptr);

--- a/Detectors/TPC/simulation/macro/toyCluster.C
+++ b/Detectors/TPC/simulation/macro/toyCluster.C
@@ -58,6 +58,7 @@
 #include "TPCReconstruction/HwClusterer.h"
 #include "TPCSimulation/GEMAmplification.h"
 #endif
+#include <boost/property_tree/ptree.hpp>
 
 using namespace o2::tpc;
 void fillTPCHits(const float theta, const float phi, const float dedx, std::vector<HitGroup>& hitGroupSector, std::pair<GlobalPosition3D, GlobalPosition3D>& trackInfo);

--- a/Detectors/TPC/workflow/src/ApplyCCDBCalibSpec.cxx
+++ b/Detectors/TPC/workflow/src/ApplyCCDBCalibSpec.cxx
@@ -19,6 +19,7 @@
 #include "TPCBase/ParameterGas.h"
 #include "DataFormatsTPC/LtrCalibData.h"
 #include "TPCWorkflow/ApplyCCDBCalibSpec.h"
+#include <boost/property_tree/ptree.hpp>
 
 using namespace o2::framework;
 

--- a/macro/o2sim.C
+++ b/macro/o2sim.C
@@ -39,6 +39,7 @@
 #include <sstream>
 #endif
 #include "migrateSimFiles.C"
+#include <boost/property_tree/ptree.hpp>
 
 void check_notransport()
 {

